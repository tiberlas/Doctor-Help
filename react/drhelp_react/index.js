const express = require('express');
const app = express();

require('./routes')(app); //these are your api routes

//Non api requests in production
if (process.env.NODE_ENV === 'production' || process.env.NODE_ENV === 'staging') {
    // Add production middleware such as redirecting to https

    // Express will serve up production assets i.e. main.js
    app.use(express.static(__dirname + '/client/build'));
    // If Express doesn't recognize route serve index.html
    const path = require('path');
    app.get('*', (req, res) => {
        res.sendFile(
            path.resolve(__dirname, 'client', 'build', 'index.html')
        );
    });
}

//start server
const PORT = process.env.PORT || 5000; //Heroku sets port dynamically
app.listen(PORT, () => {
    console.log('listening...');
}).on('error', err => {
    console.log(`Error Code: ${err.code}`);
});