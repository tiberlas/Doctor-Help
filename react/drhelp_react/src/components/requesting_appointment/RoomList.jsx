import React, { Component } from 'react';
import {ClinicAdminContext} from '../../context/ClinicAdminContextProvider';
import RoomReservingItem from './RoomReservingItem.jsx';
import axios from 'axios';
import Table from '@material-ui/core/Table';
import TableBody from '@material-ui/core/TableBody';
import TableCell from '@material-ui/core/TableCell';
import TableHead from '@material-ui/core/TableHead';
import TableRow from '@material-ui/core/TableRow';
import Button from 'react-bootstrap/Button';
import RoomModalSearch from '../rooms/RoomModalSearch';

const sortTypes = {
    name_up: {
        fn: (a, b) => b.name.localeCompare(a.name)
    },
    name_down: {
        fn: (a, b) => a.name.localeCompare(b.name)
    },
    number_up: {
        fn: (a, b) => a.number - b.number
    },
    number_down: {
        fn: (a, b) => b.number - a.number
    },
    date_up: {
        fn: (a, b) => b.firstFreeSchedule.localeCompare(a.firstFreeSchedule)
    },
    date_down: {
        fn: (a, b) => a.firstFreeSchedule.localeCompare(b.firstFreeSchedule)
    },
    default: {
        fn: (a, b) => a
    }
}

class RoomList extends Component {
    state = {
        type: this.props.type,
        rooms: [],
        currentSort: 'default',
        modalShow: false,
        name: ''
    }

    static contextType = ClinicAdminContext;

    componentDidMount() {
        this.handleGetAllRooms();
    }

    componentDidUpdate(prevProps, prevState) {
        if (prevProps.type !== this.props.type) {
            this.setState({type: this.props.type}, () => {
                this.handleGetAllRooms();
            })
        }
    }

    handleGetAllRooms = () => {
        axios.get('http://localhost:8080/api/rooms/all/type='+this.state.type)
            .then(response => {
                this.setState({rooms: response.data, currentSort: 'default'});
            })
    }

    onSortChange = (name) => {
		const { currentSort } = this.state;
        let nextSort;

        if(name === 'name') {
            if (currentSort === 'name_down') nextSort = 'name_up';
            else if (currentSort === 'name_up') nextSort = 'default';
            else nextSort = 'name_down';
        } else if(name === 'date') {
            if (currentSort === 'date_down') nextSort = 'date_up';
            else if (currentSort === 'date_up') nextSort = 'default';
            else nextSort = 'date_down';
        } else {
            if (currentSort === 'number_down') nextSort = 'number_up';
            else if (currentSort === 'number_up') nextSort = 'default';
            else nextSort = 'number_down';
        }

		this.setState({
			currentSort: nextSort
		}, () => {
            this.renderArrowName();
            this.renderArrowNumber();
            this.renderArrowDate();
        });
    };

    renderArrowName = () => {
        if(this.state.currentSort === 'name_up') {
            return '\u2191'
        } else if(this.state.currentSort === 'name_down') {
            return '\u2193'
        } else {
            return ''
        }
    }

    renderArrowNumber = () => {
        if(this.state.currentSort === 'number_up') {
            return '\u2191'
        } else if(this.state.currentSort === 'number_down') {
            return '\u2193'
        } else {
            return ''
        }
    }

    renderArrowDate = () => {
        if(this.state.currentSort === 'date_up') {
            return '\u2191'
        } else if(this.state.currentSort === 'date_down') {
            return '\u2193'
        } else {
            return ''
        }
    }

    handleSearchClick = () => {
        this.setState({modalShow: !this.state.modalShow})
    }

    handleHideModal = () => {
        this.setState({modalShow: false})
    }

    handleShowAll = () => {
        this.handleGetAllRooms();
    }

    handleRoomSearch = (name, number, date, time) => {
        this.setState({modalShow: false})
        let rname = null
        let rnumber = null
        let rdate = null 
        
        if(name !== "" && name !== null) {
            rname = name
        }
        if(number !== "" && number !== null) {
            rnumber = number
        }
        if(date !== "" && date !== null && time !== "" && time !== null) {
            rdate = date + " " + time;
        }

        console.log(rname, rnumber, rdate)

        axios.post('http://localhost:8080/api/rooms/search', {
            name: rname,
            number: rnumber,
            typeId: this.state.type,
            date: rdate
        }).then((response) => {
            this.setState({rooms: response.data})//daje sa drugin nazivima treba premapirati
        })

    }

    render() {
        let i = 0;
        return (
            <div class='row d-flex justify-content-center'>
            <div class='col-md-7'> 
                <h4>List of rooms</h4>
                <br/>
                <Table class="table table-hover ">
                    <TableHead class="table-active">
                        <TableRow class="table-active">
                            <TableCell class="text-success cursor-pointer" onClick={() => this.onSortChange('name')}>room name{this.renderArrowName()}</TableCell>
                            <TableCell class="text-success cursor-pointer" onClick={() => this.onSortChange('number')}>number{this.renderArrowNumber()}</TableCell>
                            <TableCell class="text-success cursor-pointer" onClick={() => this.onSortChange('date')}>first free date{this.renderArrowDate()}</TableCell>
                            <TableCell class="text-success"><Button class="btn btn-success" onClick={this.handleShowAll} >show all</Button></TableCell>
                            <TableCell class="text-success"><Button class='btn btn-success' onClick={this.handleSearchClick}>search</Button></TableCell>
                        </TableRow>
                    </TableHead>
                    <TableBody>
                        {this.state.rooms.sort(sortTypes[this.state.currentSort].fn).map (c => (
                            <TableRow className={(++i)%2? `table-dark` : ``} >
                                <RoomReservingItem key={c.id} id={c.id} value={c} />
                            </TableRow>
                        ))  }

                    </TableBody>
                </Table>

                <RoomModalSearch
                    hasType={true}
                    show={this.state.modalShow}
                    onHide={this.handleHideModal}
                    handleSearch={(name, number, date, time) => this.handleRoomSearch(name, number, date, time)}
                />

            </div>
            </div>
        );
    }
}
 
export default RoomList;