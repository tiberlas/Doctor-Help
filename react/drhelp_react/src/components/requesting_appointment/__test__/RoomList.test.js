import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { render, fireEvent, cleanup } from "@testing-library/react";
import RoomList from "../RoomList.jsx";
import "@testing-library/jest-dom";
import { shallow } from "enzyme";

afterEach(cleanup);

test("render without crashing", () => {
	const div = document.createElement("div");

	ReactDOM.render(
		<RoomList
			type="1"
			date="01/23/2020"
			appointmentId="1"
			operationId="4"
			nurse="3"
		/>,
		div,
	);
	ReactDOM.unmountComponentAtNode(div);
});

test("matches snapshot", () => {
	const roomList = renderer
		.create(
			<RoomList
				type="1"
				date="01/23/2020"
				appointmentId="1"
				operationId="4"
				nurse="3"
			/>,
		)
		.toJSON();

	expect(roomList).toMatchSnapshot();
});

test("show modal search dialog", () => {
	const { getByTestId, getByText } = render(
		<RoomList
			type="1"
			date="01/23/2020"
			appointmentId="1"
			operationId="4"
			nurse="3"
		/>,
	);

	getByText("search").simulate("click");

	const mockCallBack = jest.fn();

	const button = shallow(
		<RoomList
			type="1"
			date="01/23/2020"
			appointmentId="1"
			operationId="4"
			nurse="3"
		/>,
	);
	button.find("search").simulate("click");

	expect(getByTestId("modalSearch").show).toBeTruthy();
});
