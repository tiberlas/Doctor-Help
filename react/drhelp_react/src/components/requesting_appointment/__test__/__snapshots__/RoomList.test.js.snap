// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div
  class="row d-flex justify-content-center"
>
  <div
    class="col-md-7"
  >
    <h4>
      List of rooms
    </h4>
    <br />
    <table
      class="table table-hover "
      className="MuiTable-root"
    >
      <thead
        class="table-active"
        className="MuiTableHead-root"
      >
        <tr
          class="table-active"
          className="MuiTableRow-root MuiTableRow-head"
        >
          <th
            aria-sort={null}
            class="text-success cursor-pointer"
            className="MuiTableCell-root MuiTableCell-head"
            onClick={[Function]}
            scope="col"
          >
            room name
            
          </th>
          <th
            aria-sort={null}
            class="text-success cursor-pointer"
            className="MuiTableCell-root MuiTableCell-head"
            onClick={[Function]}
            scope="col"
          >
            number
            
          </th>
          <th
            aria-sort={null}
            class="text-success cursor-pointer"
            className="MuiTableCell-root MuiTableCell-head"
            onClick={[Function]}
            scope="col"
          >
            first free date
            
          </th>
          <th
            aria-sort={null}
            class="text-success"
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            <button
              class="btn btn-success"
              className="btn btn-primary"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              show all
            </button>
          </th>
          <th
            aria-sort={null}
            className="MuiTableCell-root MuiTableCell-head"
            scope="col"
          >
            <button
              className="btn btn-primary"
              data-testid="searchButton"
              disabled={false}
              onClick={[Function]}
              type="button"
            >
              search
            </button>
          </th>
        </tr>
      </thead>
      <tbody
        className="MuiTableBody-root"
      />
    </table>
  </div>
</div>
`;
